core/packet.h

```diff
@@ -0,0 +1,274 @@
+#ifndef BESS_PACKET_H_
+#define BESS_PACKET_H_
+
+#include <cassert>
+#include <type_traits>
+
+#include <rte_config.h>
+#include <rte_mbuf.h>
+
+#include "metadata.h"
+#include "worker.h"
+
+#include "snbuf_layout.h"
+
+/* NOTE: NEVER use rte_pktmbuf_*() directly,
+ *       unless you know what you are doing */
+static_assert(SNBUF_MBUF == sizeof(struct rte_mbuf),
+              "DPDK compatibility check failed");
+static_assert(SNBUF_HEADROOM == RTE_PKTMBUF_HEADROOM,
+              "DPDK compatibility check failed");
+
+namespace bess {
+
+// should probably be a static member of Packet
+inline phys_addr_t seg_dma_addr(struct rte_mbuf *mbuf) {
+  return mbuf->buf_physaddr + mbuf->data_off;
+}
+
+class Packet;
+typedef Packet **PacketArray;
+
+static inline Packet *__packet_alloc_pool(struct rte_mempool *pool) {
+  struct rte_mbuf *mbuf;
+
+  mbuf = rte_pktmbuf_alloc(pool);
+
+  return reinterpret_cast<Packet *>(mbuf);
+}
+
+static inline Packet *__packet_alloc() {
+  return reinterpret_cast<Packet *>(rte_pktmbuf_alloc(ctx.pframe_pool()));
+}
+
+struct rte_mempool *get_pframe_pool();
+struct rte_mempool *get_pframe_pool_socket(int socket);
+
+void init_mempool(void);
+void close_mempool(void);
+
+// For the layout of snbuf, see snbuf_layout.h
+class Packet {
```

Comment:
Consider explicitly deleting constructor/destructor, since we don't want them to be used in any case.